statesman-server/src/main/java/io/appform/statesman/server/dao/message/MessageConfigStoreCommand.java

```diff
@@ -0,0 +1,73 @@
+package io.appform.statesman.server.dao.message;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import io.appform.dropwizard.sharding.dao.LookupDao;
+import io.appform.statesman.engine.MessageConfigStore;
+import io.appform.statesman.model.MessageConfig;
+import io.appform.statesman.model.exception.ResponseCode;
+import io.appform.statesman.model.exception.StatesmanError;
+import io.appform.statesman.server.utils.MapperUtils;
+import io.appform.statesman.server.utils.WorkflowUtils;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@Singleton
+public class MessageConfigStoreCommand implements MessageConfigStore {
+
+    private final LookupDao<StoredMessageConfig> messageLookupDao;
+    private final LoadingCache<String, Optional<MessageConfig>> messageConfigCache;
+
+    @Inject
+    public MessageConfigStoreCommand(LookupDao<StoredMessageConfig> messageLookupDao) {
+        this.messageLookupDao = messageLookupDao;
+        messageConfigCache = Caffeine.newBuilder()
+                .maximumSize(1_000)
+                .expireAfterWrite(300, TimeUnit.SECONDS)
+                .refreshAfterWrite(60, TimeUnit.SECONDS)
+                .build(key -> {
+                    log.debug("Loading data for action for key: {}", key);
+                    return getFromDb(key);
+                });
+    }
+
+    private Optional<MessageConfig> getFromDb(String messageId) {
+        try {
+            return messageLookupDao.get(messageId)
+                    .map(config ->
+                            new MessageConfig(config.getMessageId(), MapperUtils.readTree(config.getMessageConfigBody())));
+        } catch (Exception e) {
+            throw StatesmanError.propagate(e, ResponseCode.DAO_ERROR);
+        }
+    }
+
+
+    @Override
+    public Optional<MessageConfig> create(MessageConfig messageConfig) {
+        try {
+            return messageLookupDao
+                    .save(WorkflowUtils.toDao(messageConfig))
+                    .map(config ->
+                            new MessageConfig(config.getMessageId(), MapperUtils.readTree(config.getMessageConfigBody())));
+
+        } catch (Exception e) {
+            throw StatesmanError.propagate(e, ResponseCode.DAO_ERROR);
+        }
+    }
+
+    @Override
+    public Optional<MessageConfig> get(String messageConfigId) {
+        try {
+            return messageConfigCache.get(messageConfigId);
```

Comment:
how are we pre populating data ? should we fall back to DB if cache miss occurs ? because ideally we don't expect to query for a msg id which doesn't exist. so either msg is present in cache or in DB. 

or are we pre populating the DB somehow ? in that case how will new writes get cached ?

Reply:
We have this method called getFromDb which is hooked into caffine while building the cache, so any cache miss will lead to a DB call.